pep-9999.rst

```diff
@@ -0,0 +1,1228 @@
+
+For example, Fedora and Debian (and their derivatives) both
+implement this split by using ``/usr/local`` for
+locally-installed packages and ``/usr`` for
+distribution-installed packages. Fedora uses
+``/usr/local/lib/python3.x/site-packages`` vs.
+``/usr/lib/python3.x/site-packages``. (Debian uses
+``/usr/local/lib/python3/dist-packages`` vs.
+``/usr/lib/python3/dist-packages`` as an additional layer of
+separation from a locally-compiled Python interpreter: if
+you build and install upstream CPython in
+``/usr/local/bin``, it will look at
+``/usr/local/lib/python3/site-packages``, and Debian wishes
+to make sure that packages installed via the locally-built
+interpreter don't show up on ``sys.path`` for the
+distribution interpreter.)
+
+Note that the ``/usr/local`` vs. ``/usr`` split is analogous
+to how the ``PATH`` environment variable typically includes
+``/usr/local/bin:/usr/bin`` and non-distribution software
+installs to ``/usr/local`` by default. This split is
+`recommended by the Filesystem Hierarchy Standard`__.
+
+.. __: https://refspecs.linuxfoundation.org/FHS_3.0/fhs/ch04s09.html
+
+There are two ways you could do this. One is, if you are
+building and packaging Python libraries directly (e.g., your
+packaging helpers unpack a PEP 517-built wheel or call
+``setup.py install``), arrange for those tools to use a
+directory that is not in a ``sysconfig`` scheme but is still
+on ``sys.path``.
+
+The other is to arrange for the default ``sysconfig`` scheme
+to change when running inside a package build versus when
+running on an installed system. The ``sysconfig``
+customization hooks from bpo-43976_ should make this easy:
+make your packaging tool set an environment variable or some
+other detectable configuration, and define a
+``get_preferred_schemes`` function to return a different
+scheme when called from inside a package build. Then you can
+use ``pip install`` as part of your distribution packaging.
+
+.. _bpo-43976: https://bugs.python.org/issue43976
+
+We propose adding a ``--scheme=...`` option to instruct pip
+to run against a specific scheme. (See `Implementation
+Notes`_ below for how pip currently determines schemes.)
+Once that's available, for local testing and possibly for
+actual packaging, you would be able to run something like
+``pip install --scheme=posix_distro`` to explicitly install
+a package into your distribution's location (bypassing
+``get_preferred_schemes``). One could also, if absolutely
+needed, use ``pip uninstall --scheme=posix_distro`` to use
+pip to remove packages from the system-managed directory,
+which addresses the (hopefully theoretical) regression in
+use case 5 in Rationale_.
+
+To install packages with pip, you would also need to either
+suppress the ``EXTERNALLY-MANAGED`` marker file to allow pip
+to run or to override it on the command line.  You may want
+to use the same means for suppressing the marker file in
+build chroots as you do in container images.
```

Comment:
@achernya pointed out to me that this is probably going to be very awkward for build chroots. (I was mostly imagining the Debian case where you can solve this with a package that dpkg-diverts the file and install that in both Docker containers and build chroots... but even there, if it's a build-dependency it's too easy to install that on a normal system.)

Here's another approach: if we extend the EXTERNALLY-MANAGED file to have a header `Permitted-Scheme: deb_system` (or whatever you name your distro scheme), then we can say that `pip install --scheme deb_system` is allowed to work without any further overrides. The semantics are that if pip is trying to install _and_ it is using any scheme other than `Permitted-Scheme`, then it's an error.

I'm not sure if we should also permit `pip install --target=/some/path` . I think the answer might be yes, so the logic would become:
* if you're installing to distutils scheme X, then look for the EXTERNALLY-MANAGED file. If you're not using a distutils scheme, allow.
* If there's a Permitted-Scheme header and it's not X, _or_ if there's no Permitted-Scheme header, error.
* Otherwise, allow.

Moreover, since pip takes command-line options as environment variables, you can just have some build tool (or your packaging) `export PIP_SCHEME=deb_system` and third-party build scripts that internally call `pip install` will do the right thing.

The downside of this approach is, of course, that it becomes very easy to install to the systemwide layout... is that a problem? I suspect it's fine; if you're specifically typing in `deb_system` you know you're doing something unusual.

Reply:
`--target` (and `--prefix`) is basically taking the scheme’s designated path, and then swap the prefix to use the specified path, i.e.

```python
# Scheme on Linux is posix_prefix for --prefix, posix_home for --target.
paths = sysconfig.get_paths(scheme=scheme, vars={"base": new_root, "platbase": new_root})
# Install into paths["purelib"] and path["platlib"]...
```

So would it be possible to make these situations depend on the existence (and content) of `{new_root}/EXTERNALLY-MANAGED` instead?

Reply:
Yeah, a Debian package build (and presumably other distro builds) will usually build into a different `DESTDIR` (not quite prefix, in the traditional sense, because the software must still assume that `/usr` is its prefix in the final install).

For `setuptools` builds, we use `setup.py --root temporary/install/location`.
For `pip` (which we don't use in package builds, atm), I'd imagine we'd use `--root` too.

So, if purelib/platlib != the standard purelib/platlib for the default scheme, it's probably safe to install to.

@geofft: Yes, I agree that explicitly selecting `--scheme deb_system` should imply that you want to clobber things in `deb_system`

Reply:
Hmm so I wrote up https://github.com/geofft/peps/pull/6 but I missed the comment above about how `--target` and `--prefix` do involve looking up the default scheme and modifying it.

On my Debian stable box, `base` doesn't cause `stdlib` to change:
```
$ python3
Python 3.7.3 (default, Jul 25 2020, 13:03:44) 
[GCC 8.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import sysconfig
>>> sysconfig.get_paths(vars=dict(base="foo"))
{'stdlib': '/usr/lib/python3.7', 'platstdlib': '/usr/lib/python3.7', 'purelib': 'foo/lib/python3.7/site-packages', 'platlib': '/usr/lib/python3.7/site-packages', 'include': '/usr/include/python3.7m', 'platinclude': '/usr/include/python3.7m', 'scripts': 'foo/bin', 'data': 'foo'}
```

(It does on Python 2, amusingly enough.)

So making this depend on whether the final `stdlib` path contains an `EXTERNALLY-MANAGED` file won't work right - it still will.

Should we explicitly say "and if you are not overriding variables in the scheme to move things around"? I'm not quite sure how to phrase this.

I suppose the answer is some phrasing to the effect of "If you're using an _unmodified_ distutils scheme" but also phrasing it in a way where `--root` doesn't count as modifying it. That is, even though `pip install --root $DESTDIR` installs to `$DESTDIR/usr/local/lib/python3.x/site-packages` instead of `/usr/local/lib/python3.x/site-packages`, it should still get you an error. (As a concrete example, there's no real difference between an FPM-packaged thing in /usr/local/lib and a directly-installed-by-the-sysadmin thing. If you're going to be installing Python packages with FPM, you still need to put them in a virtualenv, and maybe you want `dh_virtualenv`.)

Reply:
> On my Debian stable box, base doesn't cause stdlib to change:

Because it uses `installed_base`:
```py
    'posix_prefix': {
        'stdlib': '{installed_base}/{platlibdir}/python{py_version_short}',
        'platstdlib': '{platbase}/{platlibdir}/python{py_version_short}',
        'purelib': '{base}/lib/python{py_version_short}/site-packages',
        'platlib': '{platbase}/{platlibdir}/python{py_version_short}/site-packages',
        'include':
            '{installed_base}/include/python{py_version_short}{abiflags}',
        'platinclude':
            '{installed_platbase}/include/python{py_version_short}{abiflags}',
        'scripts': '{base}/bin',
        'data': '{base}',
        },
```

> and if you are not overriding variables in the scheme to move things around"?

Yeah, the goal is to be sure that we're installing off of the path that a default system python will use.

But it's hard to know what those `sys.path` components will be. Debian's `site.py` inserts some things, probably more than other distros do. I can't think of a foolproof way to know what values are, without firing up a child `/usr/bin/python` in a controlled environment (no `PYTHONPATH`, `PYTHONHOME`, etc.). Although even that could miss directories that would be on `sys.path` if they existed. Maybe explicitly listing them in EXTERNALLY-MANAGED is the only way?

Reply:
Yes, sorry, my code example above was only illustrative and not correct, because I didn’t want the giant `base` `installed_base` `platbase` etc. dict to obscure the point I wanted to make. I should’ve pointed this out in the comment.